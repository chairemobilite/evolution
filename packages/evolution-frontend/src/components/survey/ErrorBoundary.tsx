/*
 * Copyright 2024, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import SurveyErrorPage from '../pages/SurveyErrorPage';
import { connect } from 'react-redux';
import { UserInterviewAttributes } from 'evolution-common/lib/services/interviews/interview';
import { reportClientSideException } from '../../services/errorManagement/errorHandling';

interface ErrorProps {
    // No props required
    interview?: UserInterviewAttributes;
}

interface ErrorState {
    hasError: boolean;
}

/**
 * Error boundary component to catch exceptions in react components and display
 * a fallback UI.
 */
export class ErrorBoundary extends React.Component<React.PropsWithChildren<ErrorProps>, ErrorState> {
    constructor(props: React.PropsWithChildren<ErrorProps>) {
        super(props);
        this.state = { hasError: false };
    }

    componentDidCatch(error: Error, info: React.ErrorInfo): void {
        // Display fallback UI
        this.setState({ hasError: true });
        console.log('An exception occurred in a react component', error, info);
        // Send update responses to the server
        reportClientSideException(error, this.props.interview?.id);
    }

    resetErrorBoundary = () => {
        this.setState({ hasError: false });
    };

    render(): React.ReactNode {
        if (this.state.hasError) {
            // Go to error page as fallback UI
            return <SurveyErrorPage onRedirect={this.resetErrorBoundary} />;
        }
        return this.props.children;
    }
}

const mapStateToProps = (state, props) => {
    return {
        interview: state.survey.interview
    };
};

export default connect(mapStateToProps)(ErrorBoundary);
