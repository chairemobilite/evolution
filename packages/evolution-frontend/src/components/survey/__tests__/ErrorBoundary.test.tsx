/*
 * Copyright 2024, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import _cloneDeep from 'lodash/cloneDeep';
import React from 'react';
import TestRenderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';

import ErrorBoundary from '../ErrorBoundary';

beforeEach(() => {
    jest.clearAllMocks();
});

test('Render child widget if no error', () => {
    const NormalComponent = () => <div>Normal component</div>;
    const wrapper = TestRenderer.create(
        <MemoryRouter>
            <ErrorBoundary>
                <NormalComponent />
            </ErrorBoundary>
        </MemoryRouter>
    );
    expect(wrapper).toMatchSnapshot();
});

test('Render error page if widget has error', () => {
    const ErrorComponent = () => {
        throw new Error('Error component');
    };
    const wrapper = TestRenderer.create(
        <MemoryRouter>
            <ErrorBoundary>
                <ErrorComponent />
            </ErrorBoundary>
        </MemoryRouter>
    );
    expect(wrapper).toMatchSnapshot();
});

test('Reset the error boundary when clicking', () => {
    let hasThrown = false
    const ErrorComponent = () => {
        if (!hasThrown) {
            hasThrown = true;
            throw new Error('Error component');
        }
        return <div>Correct component</div>;
    };

    // Add initial keys for snapshots to work with enzyme without random keys (https://github.com/remix-run/react-router/issues/5579)
    const surveyErrorPage = mount(<MemoryRouter initialEntries={[ { pathname: '/', key: 'testKey' } ]}>
        <ErrorBoundary>
            <ErrorComponent />
        </ErrorBoundary>
    </MemoryRouter>);

    // This snapshot should be the error page
    expect(surveyErrorPage).toMatchSnapshot();

    // Find and click on the link in the error page
    const link = surveyErrorPage.find('a');
    link.simulate('click');

    // The second snapshot should be the correct component reset by the error boundary
    surveyErrorPage.update();
    expect(surveyErrorPage).toMatchSnapshot();
});
