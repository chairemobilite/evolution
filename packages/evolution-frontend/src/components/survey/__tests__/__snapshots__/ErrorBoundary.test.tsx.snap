// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render child widget if no error 1`] = `
<div>
  Normal component
</div>
`;

exports[`Render error page if widget has error 1`] = `
<div
  className="survey"
  id="surveyErrorPage"
  style={
    {
      "display": "block",
    }
  }
>
  <div
    className="apptr__form"
  >
    <p
      className="_large _strong _blue"
    >
      <div
        dangerouslySetInnerHTML={
          {
            "__html": "AnErrorOccurred",
          }
        }
      />
    </p>
  </div>
  <div
    className="apptr__separator"
  />
  <div
    className="apptr__form"
  >
    <div
      dangerouslySetInnerHTML={
        {
          "__html": "MakeSureReliableConnection",
        }
      }
    />
  </div>
  <div
    className="apptr__separator"
  />
  <div
    className="apptr__form"
  >
    <div
      dangerouslySetInnerHTML={
        {
          "__html": "ErrorPersistsWhatToDo",
        }
      }
    />
  </div>
  <div
    className="apptr__separator"
  />
  <div
    className="apptr__form"
  >
    <a
      href="/survey"
      onClick={[Function]}
    >
      BackToSurveyPage
    </a>
  </div>
</div>
`;

exports[`Reset the error boundary when clicking 1`] = `
<Provider
  store={
    {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <MemoryRouter
    initialEntries={
      [
        {
          "key": "testKey",
          "pathname": "/",
        },
      ]
    }
  >
    <Router
      history={
        {
          "action": "POP",
          "block": [Function],
          "canGo": [Function],
          "createHref": [Function],
          "entries": [
            {
              "hash": "",
              "key": "testKey",
              "pathname": "/",
              "search": "",
            },
          ],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "index": 0,
          "length": 1,
          "listen": [Function],
          "location": {
            "hash": "",
            "key": "testKey",
            "pathname": "/",
            "search": "",
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Connect(ErrorBoundary)>
        <ErrorBoundary
          dispatch={[Function]}
          interview={
            {
              "id": 1,
            }
          }
        >
          <withI18nextTranslation(SurveyErrorPage)
            onRedirect={[Function]}
          >
            <SurveyErrorPage
              i18n={
                {
                  "language": "en",
                  "languages": [
                    "en",
                  ],
                  "options": {
                    "appendNamespaceToCIMode": false,
                    "appendNamespaceToMissingKey": false,
                    "contextSeparator": "_",
                    "debug": false,
                    "defaultNS": "translations",
                    "fallbackLng": [
                      "en",
                    ],
                    "fallbackNS": false,
                    "ignoreJSONStructure": true,
                    "initAsync": true,
                    "interpolation": {
                      "escapeValue": false,
                      "format": [Function],
                      "formatSeparator": ",",
                      "maxReplaces": 1000,
                      "nestingOptionsSeparator": ",",
                      "nestingPrefix": "$t(",
                      "nestingSuffix": ")",
                      "prefix": "{{",
                      "skipOnVariables": true,
                      "suffix": "}}",
                      "unescapePrefix": "-",
                    },
                    "joinArrays": false,
                    "keySeparator": ".",
                    "lng": "en",
                    "load": "all",
                    "missingInterpolationHandler": false,
                    "missingKeyHandler": false,
                    "nonExplicitSupportedLngs": false,
                    "ns": [
                      "translations",
                    ],
                    "nsSeparator": ":",
                    "overloadTranslationOptionHandler": [Function],
                    "parseMissingKeyHandler": false,
                    "partialBundledLanguages": false,
                    "pluralSeparator": "_",
                    "postProcess": false,
                    "postProcessPassResolved": false,
                    "preload": false,
                    "resources": {
                      "en": {
                        "translations": {},
                      },
                    },
                    "returnEmptyString": true,
                    "returnNull": false,
                    "returnObjects": false,
                    "returnedObjectHandler": false,
                    "saveMissing": false,
                    "saveMissingPlurals": true,
                    "saveMissingTo": "fallback",
                    "simplifyPluralSuffix": true,
                    "supportedLngs": false,
                    "updateMissing": false,
                  },
                  "resolvedLanguage": undefined,
                  "store": {
                    "en": {
                      "translations": {},
                    },
                  },
                }
              }
              onRedirect={[Function]}
              t={[Function]}
              tReady={true}
            >
              <div
                className="survey"
                id="surveyErrorPage"
                style={
                  {
                    "display": "block",
                  }
                }
              >
                <div
                  className="apptr__form"
                >
                  <p
                    className="_large _strong _blue"
                  >
                    <div
                      dangerouslySetInnerHTML={
                        {
                          "__html": "AnErrorOccurred",
                        }
                      }
                    />
                  </p>
                </div>
                <div
                  className="apptr__separator"
                />
                <div
                  className="apptr__form"
                >
                  <div
                    dangerouslySetInnerHTML={
                      {
                        "__html": "MakeSureReliableConnection",
                      }
                    }
                  />
                </div>
                <div
                  className="apptr__separator"
                />
                <div
                  className="apptr__form"
                >
                  <div
                    dangerouslySetInnerHTML={
                      {
                        "__html": "ErrorPersistsWhatToDo",
                      }
                    }
                  />
                </div>
                <div
                  className="apptr__separator"
                />
                <div
                  className="apptr__form"
                >
                  <Link
                    onClick={[Function]}
                    to="/survey"
                  >
                    <LinkAnchor
                      href="/survey"
                      navigate={[Function]}
                      onClick={[Function]}
                    >
                      <a
                        href="/survey"
                        onClick={[Function]}
                      >
                        BackToSurveyPage
                      </a>
                    </LinkAnchor>
                  </Link>
                </div>
              </div>
            </SurveyErrorPage>
          </withI18nextTranslation(SurveyErrorPage)>
        </ErrorBoundary>
      </Connect(ErrorBoundary)>
    </Router>
  </MemoryRouter>
</Provider>
`;

exports[`Reset the error boundary when clicking 2`] = `
<Provider
  store={
    {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <MemoryRouter
    initialEntries={
      [
        {
          "key": "testKey",
          "pathname": "/",
        },
      ]
    }
  >
    <Router
      history={
        {
          "action": "POP",
          "block": [Function],
          "canGo": [Function],
          "createHref": [Function],
          "entries": [
            {
              "hash": "",
              "key": "testKey",
              "pathname": "/",
              "search": "",
            },
          ],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "index": 0,
          "length": 1,
          "listen": [Function],
          "location": {
            "hash": "",
            "key": "testKey",
            "pathname": "/",
            "search": "",
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Connect(ErrorBoundary)>
        <ErrorBoundary
          dispatch={[Function]}
          interview={
            {
              "id": 1,
            }
          }
        >
          <ErrorComponent>
            <div>
              Correct component
            </div>
          </ErrorComponent>
        </ErrorBoundary>
      </Connect(ErrorBoundary)>
    </Router>
  </MemoryRouter>
</Provider>
`;
